src/aspire/utils/BlockDiagonal.pyx

```diff
@@ -0,0 +1,561 @@
+        """
+        if not isinstance(other, BlockDiagonal):
+            raise RuntimeError(
+                "Attempt BlockDiagonal Matrix multiplication "
+                "(matmul,@) of non BlockDiagonal {}, try (*,mul)".format(
+                    repr(other)))
+
+        self.check_compatible(other)
+
+        C = BlockDiagonal(self.nblocks, dtype=self.dtype)
+        # note, we can do this in parallel
+        for i in range(self.nblocks):
+            C[i] = self[i] @ other[i]
+        return C
+
+    def __matmul__(self, other):
+        return self.matmul(other)
+
+    def __imatmul__(self, other):
+        self.check_compatible(other)
+        for i in range(self.nblocks):
+            self[i] @= other[i]
+
+        return self
+
+    def mul(self, val):
+        """
+        Compute the numeric (elementwise) multiplication of a BlockDiagonal matrix and a scalar.
+
+        :param other: The rhs BlockDiagonal matrix
+        :return: A BlockDiagonal of self * other.
+        """
+        # TODO, probably I should only handle a few specfic instances instead...
+        if isinstance(val, BlockDiagonal):
+            raise RuntimeError("Attempt numeric multiplication (*,mul) of two BlockDiagonals, try (matmul,@)")
+
+        C = BlockDiagonal(self.nblocks, dtype=self.dtype)
+        # note, we can do this in parallel
+        for i in range(self.nblocks):
+            C[i] = self[i] * val
+        return C
+
+    def __mul__(self, val):
+        return self.mul(val)
+
+    def __imul__(self, val):
+        if isinstance(val, BlockDiagonal):
+            raise RuntimeError("Attempt numeric multiplication (*,mul) of two BlockDiagonals, try (matmul,@)")
+        for i in range(self.nblocks):
+            self[i] *= val
+
+        return self
+
+    def __rmul__(self, other):
+        """ Convenience function, elementwise rmul commutes to mul """
+        return self.mul(other)
+
+    def __neg__(self):
+        """
+        Compute the unary negation of BlockDiagonal matrix.
+
+        :return: A BlockDiagonal like self
+        """
+        C = BlockDiagonal(self.nblocks, dtype=self.dtype)
+        for i in range(self.nblocks):
+            C[i] = -self[i]
+        return C
+
+    def __abs__(self):
+        """
+        Compute the elementwise absolute value of BlockDiagonal matrix.
+
+        :return: A BlockDiagonal like self
+        """
+        C = BlockDiagonal(self.nblocks, dtype=self.dtype)
+        for i in range(self.nblocks):
+            C[i] = np.abs(self[i])
+        return C
+
+    def norm(self, order=2):
+        """
+        Compute the norm of a BlockDiagonal matrix.
+
+        :param order: Norm order, see np.norm. Defaults to order 2 norm.
+        :return: The norm of the BlockDiagonal matrix
+        """
+        IMPLEMENTED_ORDERS  = (2,)
+        if order not in IMPLEMENTED_ORDERS:
+            raise NotImplementedError("Order {} not yet implemented, only {}".format(ord, IMPLEMENTED_ORDERS))
```

Comment:
Try to keep the number of  characters in a line to 76.

Reply:
No problem, sometimes I leave them if they are totally trivial, since its 2020, but I can be more militant about it and other pep8 style.